File: Session.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tic_Tac_Toe
{
    class Session : ISession
    {
        private Board board;
        private int boardSize;
        private char currentPlayer;
        private char player1Symbol;
        private char player2Symbol;

        public Session(int boardSize, char player1Symbol, char player2Symbol)
        {
            this.boardSize = boardSize;
            board = new Board(boardSize);
            this.player1Symbol = player1Symbol;
            this.player2Symbol = player2Symbol;
            currentPlayer = player1Symbol;
        }

        public void StartGame()
        {
            do
            {
                board.PrintBoard();
                MakeMove();
                SwitchPlayer();
            } while (!CheckGameOver());

            board.PrintBoard();
            if (currentPlayer == player1Symbol) 
            {
                Console.WriteLine($"Player {player2Symbol} wins!");
            }
            else
            {
                {
                    Console.WriteLine($"Player {player1Symbol} wins!");
                }
            } 
        }

        private void MakeMove()
        {
            int row, col;
            do
            {
                Console.WriteLine($"Player {currentPlayer}, enter your move (row and column, separated by space):");
                string[] input = Console.ReadLine().Split(' ');

                if (input.Length != 2 || !int.TryParse(input[0], out row) || !int.TryParse(input[1], out col))
                {
                    Console.WriteLine("Invalid input. Try again.");
                    continue;
                }

                row--;
                col--;

                if (!board.IsValidMove(row, col))
                {
                    Console.WriteLine("Invalid move. Try again.");
                    continue;
                }

                board.MakeMove(row, col, currentPlayer);
                break;
            } while (true);
        }

        private void SwitchPlayer()
        {
            currentPlayer = (currentPlayer == player1Symbol) ? player2Symbol : player1Symbol;
        }

        private bool CheckGameOver()
        {
            if (board.CheckWinner(player1Symbol) || board.CheckWinner(player2Symbol))
                return true;
            for (int i = 0; i < boardSize; i++)
            {
                for (int j = 0; j < boardSize; j++)
                {
                    if (board.IsValidMove(i, j))
                        return false;
                }
            }

            Console.WriteLine("It's a tie!");
            return true;
        }
    }
}

File: GameBot/TicTacToe/Session.py
from TicTacToe.Board import Board


class Session:
  def __init__(self, board_size, player1_symbol, player2_symbol, bot, message):
    self.board_size = board_size
    self.bot = bot
    self.message = message
    self.board = Board(board_size, bot, message)
    self.player1_symbol = player1_symbol
    self.player2_symbol = player2_symbol
    self.current_player = player1_symbol

  def start_game(self, message):
    while not self.check_game_over():
      self.board.print_board(message)
      self.bot.send_message(message.chat.id, f"Player {self.current_player}, enter your move (row and column, separated by space): ")
      self.make_move(message)
      self.switch_player()

    self.board.print_board()
    self.bot.send_message(self.message.chat.id, f"Player {self.current_player} wins!")

  def make_move(self, message):
    while True:
        lst = list(map(int, message.text.split()))
        row = lst[0], col = lst[1]
        if not self.board.is_valid_move(row, col):
            self.bot.send_message(message.chat.id, "Invalid move. Try again.")
            continue

        self.board.make_move(row, col, self.current_player)
        break

  def switch_player(self):
    self.current_player = self.player2_symbol if self.current_player == self.player1_symbol else self.player1_symbol

  def check_game_over(self):
    if self.board.check_winner(self.player1_symbol) or self.board.check_winner(self.player2_symbol):
      return True

    for i in range(self.board_size):
      for j in range(self.board_size):
        if self.board.is_valid_move(i, j):
          return False

    self.bot.send_message(self.message.chat.id, "It's a tie!")
    return True
